pkg/server/container_update_resources.go

```diff
@@ -30,90 +30,122 @@ import (
+	"github.com/containerd/cri/pkg/store/sandbox"
+	sandboxstore "github.com/containerd/cri/pkg/store/sandbox"
 	"github.com/containerd/cri/pkg/util"
 )
 
 // UpdateContainerResources updates ContainerConfig of the container.
 func (c *criService) UpdateContainerResources(ctx context.Context, r *runtime.UpdateContainerResourcesRequest) (retRes *runtime.UpdateContainerResourcesResponse, retErr error) {
-	container, err := c.containerStore.Get(r.GetContainerId())
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to find container")
+	if err := c.genericUpdateContainerResources(ctx, r.GetContainerId(), r); err != nil {
+		return nil, err
 	}
+	return &runtime.UpdateContainerResourcesResponse{}, nil
+}
+
+// genericUpdateContainerResources is a helper function that is capable of handling resources from UpdateContainerResources
+// and UpdateContainerResourcesV2
+func (c *criService) genericUpdateContainerResources(ctx context.Context, id string, r *runtime.UpdateContainerResourcesRequest) (retErr error) {
 	// Update resources in status update transaction, so that:
 	// 1) There won't be race condition with container start.
 	// 2) There won't be concurrent resource update to the same container.
-	if err := container.Status.Update(func(status containerstore.Status) (containerstore.Status, error) {
-		return status, c.updateContainerResources(ctx, container, r.GetLinux(), status)
-	}); err != nil {
-		return nil, errors.Wrap(err, "failed to update resources")
+	cntr, err := c.containerStore.Get(id)
+	if err == nil {
+		if err := cntr.Status.Update(func(status containerstore.Status) (containerstore.Status, error) {
+			id := cntr.ID
+			// Do not update the container when there is a removal in progress.
+			if status.Removing {
+				return status, errors.Errorf("container %q is in removing state", id)
+			}
+
+			// Update container spec. If the container is not started yet, updating
+			// spec makes sure that the resource limits are correct when start;
+			// if the container is already started, updating spec is still required,
+			// the spec will become our source of truth for resource limits.
+			return status, c.updateContainerResourcesV2(ctx, cntr.Container, r)
+		}); err != nil {
+			return errors.Wrap(err, "failed to update resources")
+		}
+	} else if err == store.ErrNotExist {
+		sndbx, err := c.sandboxStore.Get(id)
+		if err != nil {
+			return errors.Wrap(err, "failed to find container or sandbox")
+		}
+		if err := sndbx.Status.Update(func(status sandboxstore.Status) (sandboxstore.Status, error) {
+			id := cntr.ID
+			if status.State == sandbox.StateUnknown || status.State == sandbox.StateNotReady {
+				return status, errors.Errorf("sandbox %q must be running to update resources, instead has state %d", id, status.State)
+			}
+
+			return status, c.updateContainerResourcesV2(ctx, sndbx.Container, r)
+		}); err != nil {
+			return errors.Wrap(err, "failed to update resources")
+		}
+
+	} else if err != nil {
```

Comment:
Nit: redundant nil check as if we got here the error can't be nil

Reply:
Done